stateproof/builder.go

```diff
@@ -20,48 +20,175 @@ import (
+func (spw *Worker) OnPrepareVoterCommit(oldBase basics.Round, newBase basics.Round, votersFetcher ledgercore.LedgerForSPBuilder) {
+	for rnd := oldBase + 1; rnd <= newBase; rnd++ {
+		header, err := votersFetcher.BlockHdr(rnd)
+		if err != nil {
+			spw.log.Errorf("OnPrepareVoterCommit(%d): could not fetch round header: %v", rnd, err)
+			continue
+		}
+
+		proto := config.Consensus[header.CurrentProtocol]
+		if proto.StateProofInterval == 0 || uint64(rnd)%proto.StateProofInterval != 0 {
+			continue
+		}
+
+		var builderExists bool
+		err = spw.db.Atomic(func(ctx context.Context, tx *sql.Tx) (err error) {
+			builderExists, err = builderExistInDB(tx, rnd)
+			return err
+		})
+		if err != nil {
+			spw.log.Warnf("OnPrepareVoterCommit(%d): could not check builder existence, assuming it doesn't exist: %v\n", rnd, err)
+		} else if builderExists {
+			continue
+		}
+
+		buildr, err := createBuilder(rnd, votersFetcher)
+		if err != nil {
+			if errors.Is(err, errVotersNotTracked) {
+				// There are few reasons why we might encounter a situation where we don't
+				// have voters for a state proof round.
+				//
+				// 1 - When state proof chain starts, the first round s.t round  % proto.stateproofInterval == 0  will not
+				// have voters (since they are not enable). For this round we will not create a state proof.
+				// e.g if  proto.stateproofInterval == 10, and round = 10. We skip the state proof for that round
+				// (since there are not voters on round 0)
+				//
+				// 2 - When a node uses fastcatchup to some round, and immediately tries to create a builder.
+				// Node might fail to create the builder since MaxBalLookback (in catchpoint) might not be large enough
+				spw.log.Warnf("OnPrepareVoterCommit(%d): %v", rnd, err)
+				continue
+			}
+
+			spw.log.Errorf("OnPrepareVoterCommit(%d): could not create builder: %v", rnd, err)
+			continue
+		}
+
+		// At this point, there is a possibility that the signer has already created this specific builder
+		// (signer created  the builder after builderExistInDB was called and was fast enough to persist it).
+		// In this case we will rewrite the new builder
+		err = spw.db.Atomic(func(_ context.Context, tx *sql.Tx) error {
+			return persistBuilder(tx, rnd, &buildr)
+		})
+		if err != nil {
+			spw.log.Errorf("OnPrepareVoterCommit(%d): could not persist builder: %v", rnd, err)
+		}
+	}
+}
+
+// loadOrCreateBuilderWithSignatures either loads a builder from the DB or creates a new builder.
+// this function fills the builder with all the available signatures
+func (spw *Worker) loadOrCreateBuilderWithSignatures(rnd basics.Round) (builder, error) {
+	b, err := spw.loadOrCreateBuilder(rnd)
 	if err != nil {
 		return builder{}, err
 	}
 
-	hdrProto := config.Consensus[hdr.CurrentProtocol]
-	votersRnd := rnd.SubSaturate(basics.Round(hdrProto.StateProofInterval))
-	votersHdr, err := l.BlockHdr(votersRnd)
+	if err := spw.loadSignaturesIntoBuilder(&b); err != nil {
```

Comment:
Keep in mind that in the current version, Nodes only have a signature table.
When a new version gets deployed, the node will create a builder (the builder is not loaded since the table is new and empty). Since signatures for this builder might exist, the node needs to load them as well.

Reply:
So, my observation is valid in the future, when the new code kicks in.
But, for the transition, we cannot assume that the two tables are in sync. 

If so, we can make a ticket to simplify this in the future.

Reply:
@algonautshant , can you please open an issue?